### PR TITLE
Add local repository support for tools

### PR DESCRIPTION
**Pull request by Thevakumar-Luheerathan**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/41729
Related to #40264 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java [hunk 1]
```diff
@@ -389,6 +394,7 @@ private void pushBalaToCustomRepo(Path balaFilePath) {
                 ProjectUtils.deleteDirectory(balaCachesPath);
             }
             ProjectUtils.extractBala(balaFilePath, balaDestPath);
+            handleLocalTools(balaDestPath, org, packageName, repoPath.resolve(ProjectConstants.BALA_DIR_NAME));
         } catch (IOException e) {
             throw new ProjectException("error while pushing bala file '" + balaFilePath + "' to '"
                     + ProjectConstants.LOCAL_REPOSITORY_NAME + "' repository. " + e.getMessage());
```
Comment (ShammiL) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java:400]:
```suggestion
                    + ProjectConstants.LOCAL_REPOSITORY_NAME + "' repository:  " + e.getMessage());
```

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java [hunk 2]
```diff
@@ -389,6 +394,7 @@ private void pushBalaToCustomRepo(Path balaFilePath) {
                 ProjectUtils.deleteDirectory(balaCachesPath);
             }
             ProjectUtils.extractBala(balaFilePath, balaDestPath);
+            handleLocalTools(balaDestPath, org, packageName, repoPath.resolve(ProjectConstants.BALA_DIR_NAME));
         } catch (IOException e) {
             throw new ProjectException("error while pushing bala file '" + balaFilePath + "' to '"
```
Comment (ShammiL) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java:399]: Since we are giving the `balaFilePath` here, can we improve the message? Something like "error while pushing bala at balaFilePath"

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java [hunk 3]
```diff
@@ -404,6 +410,53 @@ private void pushBalaToCustomRepo(Path balaFilePath) {
...
+            return;
+        }
+
+        JsonObject packageDesc = new JsonObject();
+        packageDesc.addProperty("org", org);
+        packageDesc.addProperty("name", packageName);
+        Path localToolJsonPath = localRepoBalaPath.resolve("local-tools.json");
+        if (localToolJsonPath.toFile().exists()) {
+            try (BufferedReader bufferedReader = Files.newBufferedReader(localToolJsonPath,
+                    StandardCharsets.UTF_8)) {
```
Comment (ShammiL) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java:439]: extra spacing can be removed

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java [hunk 4]
```diff
@@ -208,6 +225,13 @@ private void handlePullCommand() {
             return;
         }
 
+        if (ProjectConstants.LOCAL_REPOSITORY_NAME.equals(repositoryName) && EMPTY_STRING.equals(version)) {
+            CommandUtil.printError(errStream, "a version should be provided when pulling a tool from local " +
```
Comment (ShammiL) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java:229]:
```suggestion
            CommandUtil.printError(errStream, "version should be provided when pulling a tool from local " +
```

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java [hunk 5]
```diff
@@ -269,19 +293,21 @@ private void handleUseCommand() {
...
         BalToolsToml balToolsToml = BalToolsToml.from(balToolsTomlPath);
         BalToolsManifest balToolsManifest = BalToolsManifestBuilder.from(balToolsToml).build();
-        Optional<BalToolsManifest.Tool> tool = balToolsManifest.getTool(toolId, version);
+        Optional<BalToolsManifest.Tool> tool = balToolsManifest.getTool(toolId, version, repositoryName);
         if (tool.isEmpty()) {
             CommandUtil.printError(errStream, "tool '" + toolId + ":" + version + "' is not found. " +
                             "Run 'bal tool pull " + toolId + ":" + version
-                            + "' to fetch and set as the active version.", null, false);
+                            + "' or 'bal tool pull " + toolId + ":" + version
+                    + " --repository=local' to fetch and set as the active version.", null, false);
```
Comment (ShammiL) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java:301]: formatting can be improved

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java [hunk 6]
```diff
@@ -392,6 +424,12 @@ private void handleUpdateCommand() {
...
             return;
         }
+        if (ProjectConstants.LOCAL_REPOSITORY_NAME.equals(repositoryName)) {
+            CommandUtil.printError(errStream, "tool update command is not supported for local repository.",
+                    null, false);
+            CommandUtil.exitError(this.exitWhenFinish);
+            return;
+        }
 
         toolId = argList.get(1);
```
Comment (ShammiL) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java:435]:
```suggestion
```

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java [hunk 7]
```diff
@@ -430,6 +477,31 @@ public void pullToolAndUpdateBalToolsToml(String toolIdArg, String versionArg) {
...
+        JsonObject localToolJson;
+        Gson gson = new Gson();
+        Path localBalaPath = RepoUtils.createAndGetHomeReposPath().resolve(Path.of(REPOSITORIES_DIR,
+                ProjectConstants.LOCAL_REPOSITORY_NAME, BALA_DIR_NAME));
+        Path localToolJsonPath = localBalaPath.resolve("local-tools.json");
+
+        if (!Files.exists(localToolJsonPath)) {
+            return false;
+        }
+
```
Comment (ShammiL) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java:490]:
```suggestion
```

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/launcher/LauncherUtils.java [hunk 8]
```diff
@@ -138,7 +159,7 @@ private static void generateCommandDescription(CommandLine command, StringBuilde
         }
         stringBuilder.append("\n")
                 .append("        ")
```
Comment (ShammiL) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/launcher/LauncherUtils.java:161]: Can't we make use of `\t` here instead?

File: cli/ballerina-cli/src/test/java/io/ballerina/cli/cmd/PushCommandTest.java [hunk 9]
```diff
@@ -205,6 +209,59 @@ public void testPushWithCustomPath() throws IOException {
...
+        try (BufferedReader bufferedReader = Files.newBufferedReader(localToolJsonPath, StandardCharsets.UTF_8)) {
+            JsonObject localToolJson = new Gson().fromJson(bufferedReader, JsonObject.class);
+            JsonObject pkgDesc = localToolJson.get("luhee").getAsJsonObject();
+
+            Assert.assertEquals(pkgDesc.get("org").getAsString(), "gayaldassanayake");
+            Assert.assertEquals(pkgDesc.get("name").getAsString(), "tool_gayal");
+        }
+
+    }
+
```
Comment (ShammiL) [inline cli/ballerina-cli/src/test/java/io/ballerina/cli/cmd/PushCommandTest.java:264]: Noticed quite a few new lines added unnecessarily. Would be better check and remove them.

File: cli/ballerina-cli/src/test/java/io/ballerina/cli/cmd/ToolCommandTest.java [hunk 10]
```diff
@@ -33,6 +48,144 @@
...
+            toolCommand.execute();
+            buildLog = readOutput(true);
+            Assert.assertTrue(buildLog.contains("tool 'luhee:1.2.0' successfully removed.\n"));
+            balToolsToml = BalToolsToml.from(mockHomeRepo.resolve(".config").resolve("bal-tools.toml"));
+            balToolsManifest = BalToolsManifestBuilder.from(balToolsToml).build();
+            tool = balToolsManifest.getTool("luhee", "1.2.0", "local");
+            Assert.assertTrue(tool.isEmpty());
+        }
+    }
+
```
Comment (ShammiL) [inline cli/ballerina-cli/src/test/java/io/ballerina/cli/cmd/ToolCommandTest.java:188]: Constants can be used where necessary

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java [hunk 11]
```diff
@@ -269,19 +293,21 @@ private void handleUseCommand() {
...
             CommandUtil.exitError(this.exitWhenFinish);
             return;
         }
         this.org = tool.get().org();
         this.name = tool.get().name();
 
         Optional<BalToolsManifest.Tool> currentActiveTool = balToolsManifest.getActiveTool(toolId);
-        if (currentActiveTool.isPresent() && currentActiveTool.get().version().equals(tool.get().version())) {
+        if (currentActiveTool.isPresent() && currentActiveTool.get().version().equals(tool.get().version()) &&
+                Objects.equals(currentActiveTool.get().repository(), tool.get().repository())) {
```
Comment (gayaldassanayake) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java:310]: Can't we just have a string equal operation?

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java [hunk 12]
```diff
@@ -430,6 +477,31 @@ public void pullToolAndUpdateBalToolsToml(String toolIdArg, String versionArg) {
         }
     }
 
+    private boolean isToolAvailableInLocalRepo(String toolId, String version) {
+        JsonObject localToolJson;
+        Gson gson = new Gson();
+        Path localBalaPath = RepoUtils.createAndGetHomeReposPath().resolve(Path.of(REPOSITORIES_DIR,
+                ProjectConstants.LOCAL_REPOSITORY_NAME, BALA_DIR_NAME));
+        Path localToolJsonPath = localBalaPath.resolve("local-tools.json");
+
```
Comment (gayaldassanayake) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java:486]:
```suggestion
```

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java [hunk 13]
```diff
@@ -404,6 +410,53 @@ private void pushBalaToCustomRepo(Path balaFilePath) {
...
+        Gson gson = new Gson();
+        if (!balToolJsonPath.toFile().exists()) {
+            return;
+        }
+        try (BufferedReader bufferedReader = Files.newBufferedReader(balToolJsonPath, StandardCharsets.UTF_8)) {
+            balToolJson = gson.fromJson(bufferedReader, JsonObject.class);
+        } catch (IOException e) {
+            throw new ProjectException("Failed to read bal-tools.json file: " + e.getMessage());
+        }
+        toolId = balToolJson.get("tool_id").getAsString();
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java:427]: Let's introduce a constant for this.

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java [hunk 14]
```diff
@@ -404,6 +410,53 @@ private void pushBalaToCustomRepo(Path balaFilePath) {
...
+
+        if (toolId ==  null) {
+            return;
+        }
+
+        JsonObject packageDesc = new JsonObject();
+        packageDesc.addProperty("org", org);
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java:434]: Let's move this to a constant again

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java [hunk 15]
```diff
@@ -406,13 +444,22 @@ private void handleUpdateCommand() {
...
                 .resolve(REPOSITORIES_DIR).resolve(CENTRAL_REPOSITORY_CACHE_NAME)
                 .resolve(ProjectConstants.BALA_DIR_NAME);
 
         String supportedPlatform = Arrays.stream(JvmTarget.values())
                 .map(JvmTarget::code)
                 .collect(Collectors.joining(","));
+        if (ProjectConstants.LOCAL_REPOSITORY_NAME.equals(repositoryName)) {
+            if (!isToolAvailableInLocalRepo(toolId, version)) {
+                errStream.println("tool '" + toolId + ":" + version + "' is not available in local repository." +
+                        "\nPlease publish it first with 'bal push --repository=local' command.");
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java:457]:
```suggestion
                        "\nUse 'bal push --repository=local' to publish it.");
```

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/launcher/LauncherUtils.java [hunk 16]
```diff
@@ -103,13 +110,26 @@ static String generateGeneralHelp(Map<String, CommandLine> subCommands) {
         StringBuilder helpBuilder = new StringBuilder();
         helpBuilder.append(BLauncherCmd.getCommandUsageInfo(HELP));
 
+        Path balToolsTomlPath = RepoUtils.createAndGetHomeReposPath().resolve(Path.of(CONFIG_DIR, BAL_TOOLS_TOML));
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/launcher/LauncherUtils.java:113]:
```suggestion
        Path balToolsTomlPath = RepoUtils.createAndGetHomeReposPath().resolve(CONFIG_DIR).resolve( BAL_TOOLS_TOML);
```

File: cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-pull.help [hunk 17]
```diff
@@ -8,6 +8,9 @@ OPTIONS
        -h, --help
            Print the usage details of tool pull command.
 
+       --repository
+           Pull a tool from a custom repository.
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-pull.help:12]:
```suggestion
           Pull a tool from a custom repository.
```
@sameerajayasoma do we need to consider supporting custom repositories for publishing tools? If not, we can reword this to ` Pull a tool from a custom repository.`

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java (hunk 11 above)
Comment (Thevakumar-Luheerathan) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/ToolCommand.java:310]: No.. These values can be null as well.

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java [hunk 18]
```diff
@@ -404,6 +410,53 @@ private void pushBalaToCustomRepo(Path balaFilePath) {
                 + " to '" + repositoryName + "' repository.");
     }
 
+    private void handleLocalTools(Path balaDestPath, String org, String packageName, Path localRepoBalaPath) {
+        Path balToolJsonPath = balaDestPath.resolve("tool").resolve("bal-tool.json");
```
[the remark this answers is not in the document]
Comment (azinneera) [reply cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java:414]: Update other places as well.

File: cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java [hunk 19]
```diff
@@ -404,6 +410,53 @@ private void pushBalaToCustomRepo(Path balaFilePath) {
...
+        String toolId = null;
```
Comment (azinneera) [inline cli/ballerina-cli/src/main/java/io/ballerina/cli/cmd/PushCommand.java:415]: This is redundant initialization I guess.

File: cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-pull.help (hunk 17 above)
Comment (azinneera) [reply cli/ballerina-cli/src/main/resources/cli-help/ballerina-tool-pull.help:12]: Let's change this to:
```suggestion
           Pull a tool from a local repository.
```

**Review by azinneera — APPROVED** [review]
